### PR TITLE
FastMRI dataset onboarding script and detailed examples

### PR DESCRIPTION
**Pull request by ant0nsc**
This PR adds necessary tooling and examples for running fastMRI reconstruction models.
- Script to create and run an Azure Data Factory to download the raw data, and place them into a storage account
- Detailed examples to run the `VarNet` model from the fastMRI github repo
- Ability to work with fixed mounting points for datasets

### CODE REVIEW

File: docs/fastmri.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Working with FastMRI models
```
Comment (javier-alvarez) [inline docs/fastmri.md:1]: is this indexed in the README.md?

Comment (ant0nsc) [reply docs/fastmri.md:1]: it is, yes

File: .idea/runConfigurations/Template__Run_ML_on_AzureML.xml [hunk 2]
```diff
@@ -6,18 +6,19 @@
     <envs>
       <env name="PYTHONUNBUFFERED" value="1" />
     </envs>
-    <option name="SDK_HOME" value="" />
+    <option name="SDK_HOME" value="wsl://UBUNTU2004/home/antonsc/miniconda3/envs/InnerEye/bin/python" />
```
Comment (melanibe) [inline .idea/runConfigurations/Template__Run_ML_on_AzureML.xml:9]: probably not on purpose

File: InnerEye/Azure/azure_config.py [hunk 3]
```diff
@@ -242,6 +246,67 @@ def fetch_run(self, run_recovery_id: str) -> Run:
...
         return fetch_run(workspace=self.get_workspace(), run_recovery_id=run_recovery_id)
 
+    def get_or_create_dataset(self, azure_dataset_id: str) -> FileDataset:
+        """
+        Looks in the AzureML datastore for a dataset of the given name. If there is no such dataset, a dataset is
+        created and registered, assuming that the files are in a folder that has the same name as the dataset.
+        For example, if azure_dataset_id is 'foo', then the 'foo' dataset should be pointing to the folder
+        <container_root>/datasets/foo
+
+        WARNING: the behaviour of Dataset.File.from_files, used below, is idiosyncratic. For example,
```
Comment (melanibe) [inline InnerEye/Azure/azure_config.py:256]: I don't think I understand this paragraph

File: InnerEye/ML/configs/other/fastmri_varnet.py [hunk 4]
```diff
@@ -37,32 +43,131 @@ def log_image(self, name: str, image: torch.Tensor) -> None:
...
     # All fields that are declared here will be automatically available as commandline arguments.
-    challenge: str = param.String(default="multicoil", doc="Chooses between the singlecoil or multicoil"
-                                                           "acquisition setup.")
     sample_rate: Optional[float] = param.Number(default=None, doc="Fraction of slices of the training data split to "
                                                                   "use. Default: 1.0")
 
     def __init__(self) -> None:
         super().__init__()
-        self.azure_dataset_id = "fastmrimini_brain"
+        self.num_epochs = 50
```
Comment (melanibe) [inline InnerEye/ML/configs/other/fastmri_varnet.py:99]: Hard coded on purpose?

File: InnerEye/ML/configs/other/fastmri_varnet.py [hunk 5]
```diff
@@ -37,32 +43,131 @@ def log_image(self, name: str, image: torch.Tensor) -> None:
                 experiment.add_image(name, image, global_step=self.global_step)
 
 
+def get_data_module(azure_dataset_id: str,
```
Comment (melanibe) [inline InnerEye/ML/configs/other/fastmri_varnet.py:46]: Would it make sense to name that get_fastmri_data_module?

File: InnerEye/ML/deep_learning_config.py [hunk 6]
```diff
@@ -468,6 +492,10 @@ class TrainerParams(param.Parameterized):
                       doc="Controls the PyTorch Lightning trainer flags 'deterministic' and 'benchmark'. If "
                           "'pl_deterministic' is True, results are perfectly reproducible. If False, they are not, but "
                           "you may see training speed increases.")
+    pl_find_unused_parameters: bool = \
+        param.Boolean(default=False,
+                      doc="Controls the PyTorch Lightning flag 'find_unused_parameters' for the DDP plugin. "
```
Comment (melanibe) [inline InnerEye/ML/deep_learning_config.py:497]: Interesting

File: InnerEye/Scripts/prepare_fastmri.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                                                linked_service_name=blob_storage_name,
+                                                linked_service=blob_storage_service)
+
+    linked_blob_storage = LinkedServiceReference(reference_name=blob_storage_name)
+    linked_http = LinkedServiceReference(reference_name=http_name)
+
+    def download_and_uncompress(source_file_or_tuple: Union[str, Tuple[str, str]], target_folder: str) -> List[str]:
+        """
+        Downloads a file from AWS and stores them in blob storage in its compressed form.
+        Also, downloads the same file from AWS, uncompresses it, and writes the results to the given folder
```
Comment (melanibe) [inline InnerEye/Scripts/prepare_fastmri.py:166]: We download twice from AWS?

File: InnerEye/Scripts/prepare_fastmri.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    :param is_unittest: If True, download a small tar.gz file from github. If False, download the "real" fastMRI
+    datafiles from AWS.
+    :param connection_string: The connection string of the Azure storage where the downloaded data should be stored.
+    """
+
+    azure_config = AzureConfig.from_yaml(yaml_file_path=fixed_paths.SETTINGS_YAML_FILE,
+                                         project_root=fixed_paths.repository_root_directory())
+
+    # The data factory name. It must be globally unique.
+    data_factory_name = "fastmri-copy-data"
```
Comment (melanibe) [inline InnerEye/Scripts/prepare_fastmri.py:120]: Would it make sense to have this parametrizable?

File: Tests/ML/test_lightning_containers.py [hunk 9]
```diff
@@ -179,9 +192,10 @@ def test_container_to_str() -> None:
     print(s)
     assert "foo" in s
     assert "bar" in s
-    assert "param" not in s
-    assert "initialized" not in s
     assert "123456" in s
+    # These two are internal variables of the params library, and should be skipped.
+    assert "param                " not in s
```
Comment (melanibe) [inline Tests/ML/test_lightning_containers.py:197]: weird spaces on purpose?

File: InnerEye/Scripts/prepare_fastmri.py (hunk 8 above)
Comment (melanibe) [reply InnerEye/Scripts/prepare_fastmri.py:120]: ah no i see it's deleted afterwards anyways, never mind

File: docs/fastmri.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+each VM has 4 Tesla P40 cards, training will use a total of 16 GPUs.
+
+As common with multiple nodes, training time will not scale linearly with increased number of nodes. The following
+table gives a rough overview of time to train 1 epoch of the FastMri model in the InnerEye toolbox 
+on our cluster (`Standard_ND24s` nodes with 4 Tesla P40 cards):
+
+| Step | 1 node (4 GPUs) | 2 nodes (8 GPUs) | 4 nodes (16 GPUs) | 8 nodes (32 GPUs) |
+| --- | --- | --- | --- | --- |
+| Download training data (1.25 TB) | 22min | 22min | 22min | 22min |
+| Train and validate 1 epoch | 4h 15min | 2h 13min | 1h 6min | 34min |
```
Comment (melanibe) [inline docs/fastmri.md:104]: Wow 1 epoch 4hours 😮

**Review by melanibe — APPROVED** [review]
Great new capabilities added 🚀 And really liked the documentation! 
Left some minor comments, feel free to ignore if they don't make sense.

File: InnerEye/Azure/azure_config.py (hunk 3 above)
Comment (ant0nsc) [reply InnerEye/Azure/azure_config.py:256]: This is comment that David left here. I've removed it, and I think also fixed the problem he wishes to describe here.

File: InnerEye/ML/configs/other/fastmri_varnet.py (hunk 4 above)
Comment (ant0nsc) [reply InnerEye/ML/configs/other/fastmri_varnet.py:99]: yes. The FB models are all trained with 50 epochs by default.

File: InnerEye/ML/configs/other/fastmri_varnet.py (hunk 5 above)
Comment (ant0nsc) [reply InnerEye/ML/configs/other/fastmri_varnet.py:46]: yup.

File: InnerEye/ML/deep_learning_config.py (hunk 6 above)
Comment (ant0nsc) [reply InnerEye/ML/deep_learning_config.py:497]: yes :-)

File: InnerEye/Scripts/prepare_fastmri.py (hunk 7 above)
Comment (ant0nsc) [reply InnerEye/Scripts/prepare_fastmri.py:166]: Good spot. I did that initially, but that was changed. Updated comment.

File: Tests/ML/test_lightning_containers.py (hunk 9 above)
Comment (ant0nsc) [reply Tests/ML/test_lightning_containers.py:197]: yes. Adding a comment

File: .idea/runConfigurations/Template__Run_ML_on_AzureML.xml (hunk 2 above)
Comment (ant0nsc) [reply .idea/runConfigurations/Template__Run_ML_on_AzureML.xml:9]: correct, thanks!

File: .idea/runConfigurations/Template__Run_ML_on_AzureML.xml [hunk 11]
```diff
@@ -8,7 +8,7 @@
     </envs>
     <option name="SDK_HOME" value="" />
     <option name="WORKING_DIRECTORY" value="$PROJECT_DIR$/" />
-    <option name="IS_MODULE_SDK" value="false" />
+    <option name="IS_MODULE_SDK" value="true" />
```
Comment (melanibe) [inline .idea/runConfigurations/Template__Run_ML_on_AzureML.xml:11]: that on purpose?

File: InnerEye/Scripts/prepare_fastmri.py [hunk 12]
```diff
@@ -163,12 +163,13 @@ def create_datafactory_and_run(files_and_tokens: Dict[str, str],
...
-        Also, downloads the same file from AWS, uncompresses it, and writes the results to the given folder
-        in blob storage.
+        From the compressed file in blob storage, it is then uncompressed, and written to a new folder in blob storage.
+        For example, if 'target_folder' is 'foo', the uncompressed file will be written to folder 'foo', and the
+        compressed raw data will be written to 'foo_compressed'.
         :param source_file_or_tuple: The name of the .tar.gz or .tar file to download, without any access tokens.
         If the name is a Tuple[str, str], the second tuple element is the "real" extension, for files where the
         extension is misleading.
-        :param target_folder: The folder in the target storage account
+        :param target_folder: The folder prefix in the target storage account. T
```
Comment (melanibe) [inline InnerEye/Scripts/prepare_fastmri.py:172]: Little typo

Comment (melanibe) [inline InnerEye/Scripts/prepare_fastmri.py:172]:
```suggestion
        :param target_folder: The folder prefix in the target storage account.
```

**Review by melanibe — DISMISSED** [review]

**Review by melanibe — DISMISSED** [review]

**Comment by hansenms** [comment]
At the moment, the `prepare_fastmri.py` script does not work. At first glance, there are problems with binary datasets having space in their names, e.g. "<something> On Azure`. I tried a quick hack to remove all the spaces in various names, but I am now getting this error (when running the pipelines):

```
{"stack":"Error: Error: Error: TypeError: Cannot set property 'enablePartitionDiscovery' of undefined\n    at t.<anonymous> (https://ms-adf.azure.com/main.01ccec2db0cb1598b89b.js:2:1661940)\n    at https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:112098\n    at Object.throw (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:112203)\n    at s (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:110993)\n    at e.invoke (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:14216)\n    at Object.onInvoke (https://ms-adf.azure.com/main.01ccec2db0cb1598b89b.js:2:3754478)\n    at e.invoke (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:14156)\n    at t.run (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:9419)\n    at https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:21175\n    at e.invokeTask (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:14900)\n    at Object.onInvokeTask (https://ms-adf.azure.com/main.01ccec2db0cb1598b89b.js:2:3754390)\n    at e.invokeTask (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:14821)\n    at t.runTask (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:10082)\n    at g (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:17167)\n    at t.invokeTask (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:16012)\n    at invoke (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:15874)\n    at n.args.<computed> (https://ms-adf.azure.com/polyfills.7135c37b2f707063692f.js:2:45244)","message":"Error: Error: TypeError: Cannot set property 'enablePartitionDiscovery' of undefined"}
```

**Comment by hansenms** [comment]
correction on my comment above, appears there was an exit(1) in there for debugging purposes that caused the scrip to exit prematurely. Continuing testing with that fixed.

**Review by melanibe — APPROVED** [review]

File: InnerEye/ML/model_training.py [hunk 13]
```diff
@@ -152,6 +170,13 @@ def create_lightning_trainer(container: LightningContainer,
     callbacks = [best_checkpoint_callback, recovery_checkpoint_callback]
     if "callbacks" in kwargs:
         callbacks.append(kwargs.pop("callbacks"))  # type: ignore
+    is_azureml_run = not is_offline_run_context(RUN_CONTEXT)
+    progress_bar_refresh_rate = container.pl_progress_bar_refresh_rate
+    if progress_bar_refresh_rate is None and is_azureml_run:
+        # When running in AzureML, the default progress bar clutters the output files with thousands of lines.
+        progress_bar_refresh_rate = 50
```
Comment (Shruthi42) [inline InnerEye/ML/model_training.py:177]: Does this need to be set here? I imagine most containers would override this if necessary (the FastMri container does), and this introduces a small hidden difference between an offline and AzureML run.

**Review by Shruthi42 — APPROVED** [review]

**Pull request merged**